graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/CloneMessage.java

```diff
@@ -33,19 +38,48 @@ public class CloneMessage extends AbstractFunction<Message> {
 
     public static final String NAME = "clone_message";
 
+    private static final String CLONE_SOURCE = "gl2_clone_source_rule";
+    private static final String CLONE_NUMBER = "gl2_clone_number";
+    static final int MAX_CLONES = 100;
+
     private final ParameterDescriptor<Message, Message> messageParam;
+    private final ParameterDescriptor<Boolean, Boolean> loopDetectionParam;
 
     public CloneMessage() {
+        loopDetectionParam = ParameterDescriptor.bool("preventLoops").optional().description("Detects if a cloned message is processed by the same rule again, in order to prevent loops. Defaults to 'false', but will not allow more than " + MAX_CLONES + " clones if not explicitly set.").build();
         messageParam = type("message", Message.class).optional().description("The message to use, defaults to '$message'").build();
     }
 
     @Override
     public Message evaluate(FunctionArgs args, EvaluationContext context) {
         final Message currentMessage = messageParam.optional(args, context).orElse(context.currentMessage());
+        final Optional<Boolean> preventLoops = loopDetectionParam.optional(args, context);
+
+        int cloneNumber = (int) currentMessage.getMetadataValue(CLONE_NUMBER, 0);
+
+        final Rule cloneSource = (Rule) currentMessage.getMetadataValue(CLONE_SOURCE);
+        final Rule rule = context.getRule();
+        if (ObjectUtils.allNotNull(cloneSource, rule) && Objects.equals(cloneSource, rule)) { // message was cloned by the same rule before
+
+            if (preventLoops.orElse(false)) {
+                return null;
+            }
+
+            if (preventLoops.isEmpty() && cloneNumber >= MAX_CLONES) {
+                LOG.error("Message was cloned more than {} times by rule '{}'. Not allowing any more clones to prevent a potential endless loop. If this was intentional, please explicitly set the 'preventLoops' parameter to 'false'.",
+                        MAX_CLONES, cloneSource.name());
+                return null;
```

Comment:
As we consider this an error, I think it should show up in our "Indexing and Processing Errors". I think we can achieve this with something along those lines:

```suggestion
               throw new IllegalStateException(
                        StringUtils.f("Message was cloned more than %d times by rule '%s'. Not allowing any more " +
                                        "clones to prevent a potential endless loop. If this was intentional, please " +
                                        "explicitly set the 'preventLoops' parameter to 'false'.",
                                MAX_CLONES, cloneSource.name()));
```